features/user-profile.md

```diff
@@ -0,0 +1,226 @@
+- `container.discoverUserByEmail(email)`
+- `container.discoverUserByEmails(emails)`
+
+## New Sign Up and Log In API
+
+The new sign up API is proposed to enable: 1) sign up with custom data;
+2) update user profile during sign up. The new sign up API will perform:
+
+1. Discover user with the provided data
+1. Sign up with the provided data and password, if no user is found
+1. Update user record as profile
+
+The function signature will be `container.signup(authData, password, profile)`
+while `authData` is used to discover user. The following sign up API will be
+kept as helper methods:
+
+- `container.signupWithUsername(username, password)`
+- `container.signupWithEmail(email, password)`
+- `container.signupAnonymously()`
+
+The similar signature is also proposed to log in API to enable log in with
+custom data: `container.login(authData, password)`.  The new log in API will
+perform:
+
+1. Discover user with the provided data
+1. Log in with the provided data and password, if only one user is found
+
+The current log in API will be kept as helper methods:
+
+- `container.loginWithUsername(username, password)`
+- `container.loginWithEmail(email, password)`
+
+# Sample Codes for Use Cases
+
+## User discovery API
+
+The following sample code shows how to use user discovery API for different
+purposes:
+
+```js
+const skygear = require('skygear');
+
+// Use Case 1: discover user by username
+skygear.discoverUserByUsername('cheungpat')
+.then((users) => {
+  if (users.length) {
+    console.warn('Cannot find cheungpat');
+    return;
+  }
+
+  const theUser = users[0];
+  const score = theUser['game_score'];
+  console.log(`The game score of cheungpat is ${score}`);
+});
+
+// Use Case 2: discover user by discoverable fields
+skygear.discoverUser({ gender: 'male', status: 'online' })
+.then((users) => {
+  console.log(`${users.length} male users are online`);
+});
+
+// Comparison: query user
+/*
+    Since user discovery API only for equality query, for other comparisons on
+    other user fields, developers should use record query on user record.
+ */
+const User = skygear.Record.extend('user');
+const userQuery = new skygear.Query(User);
+userQuery.greaterThan('game_score', 3000);
+userQuery.equalTo('status', 'online');
+skygear.publicDB.query(userQuery)
+.then((users) => {
+  console.log(`Found ${users.length} high score users online`);
+});
+```
+
+## New sign up / log in API
+
+The usage of new sign up / log in API is shown as followed:
+
+```js
+const skygear = require('skygear');
+
+// Use Case 1: sign up and setting user profile
+skygear.signup(
+  { 'username': 'user001' }
+  's3cuRe-p@ssw0rd',
+  {
+    birthday: new Date(1992, 2, 29),
+    game_score: 1000
+  }
+).then((user) => {
+  console.log(`Successfully sign up with user ID: ${user.id}`);
+}).catch((err) => {
+  console.error(`Failed to sign up: ${err.message}`);
+});
+
+// Use Case 2: sign up with custom data
+skygear.signup(
+  { username: 'user002', role: 'doctor' },
```

Comment:
Added "Authentication Record Configurations" for setting uniqueness of the auth fields. (Better to generalize this configuration to all record, in another feature issue)